### PR TITLE
Implement read_only and read_only!

### PR DESCRIPTION
**Pull request by Joe-noh**
This implements two methods `Yao.read_only` and `Yao.read_only!` as I proposed at https://github.com/yaocloud/yao/issues/27.

I'm worried about the way to detect whether it is inside `read_only`, `read_only!` or else is not thread safe. Because it modifies and refers global setting `Yao.config`.

### CODE REVIEW

**Comment by linyows** [comment]
Great pull-request :+1:

File: lib/yao/mode.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+module Yao
+  module Mode
+    def read_only!(&blk)
+      raise unless block_given?
+
+      Yao.config.set :raise_on_write, true
+      yield(blk)
```
Comment (hanazuki) [inline lib/yao/mode.rb]: 1. This statement passes `blk` as the first argument to `blk` itself. Is it the desired behaviour? 
2. What if `yield` raise? It looks `raise_on_write` is not correctly restored to false.

File: lib/yao/mode.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      Yao.config.set :raise_on_write, false
+    end
+
+    def read_only(&blk)
+      raise unless block_given?
+
+      Yao.config.set :noop_on_write, true
+      yield(blk)
```
Comment (hanazuki) [inline lib/yao/mode.rb]: ditto here

File: lib/yao/faraday_middlewares.rb [hunk 3]
```diff
@@ -30,6 +31,39 @@ def call(env)
...
+    return @app.call(env) if allowed_request?(env)
+
+    if Yao.config.raise_on_write
+      raise Yao::ReadOnlyViolationError
+    elsif Yao.config.noop_on_write
+      env
+    else
+      @app.call(env)
+    end
+  end
```
Comment (udzura) [inline lib/yao/faraday_middlewares.rb]: This seems cool

File: lib/yao/faraday_middlewares.rb [hunk 4]
```diff
@@ -30,6 +31,39 @@ def call(env)
...
+
+  private
+
+  def allowed_request?(env)
+    return true if env[:method] == :get
+
+    allowed_requests = [
+      {method: :post, path: "/v2.0/tokens"}
+    ]
```
Comment (udzura) [inline lib/yao/faraday_middlewares.rb]: Can it be a constant?

(Maybe every time `allowed_request?` is called, hash object is generated)

File: lib/yao/mode.rb (hunk 1 above)
Comment (Joe-noh) [inline lib/yao/mode.rb]: good points. will fix.

**Comment by hanazuki** [comment]
I think the test suite should cover some reentrant conditions as the following (and the ones with combinations of `read_only` and `read_only!`)

```rb
Yao.read_only do
  assert_true Yao.config.noop_on_write
  Yao.read_only do
    assert_true Yao.config.noop_on_write
  end
  assert_true Yao.config.noop_on_write
end
```

**Comment by udzura** [comment]
Looks all good. I'd love to merge and release. cc @hanazuki

**Pull request merged**